### PR TITLE
Blobs freezer

### PR DESCRIPTION
**Pull request by emhane**
## Issue Addressed

https://github.com/sigp/lighthouse/issues/3843

## Proposed Changes

Store blobs in separate blobs freezer or in the same freezer as historical state.

## Additional Info

### CODE REVIEW

**Comment by paulhauner** [comment]
I think the implementation here looks good, however I would like to suggest a conceptual change.

In this PR the concept of "hot" and "cold" blobs is introduced. A hot blob being one from a slot *later* than the split slot (the split slot is approximates the finalized slot) and a cold blob being one from a slot *earlier* than the split slot.

For context, one of the primary reasons we have the hot/cold database separation is because the DBs have different schemas due to the following differences:

- The hot database needs to be able to handle multiple chains (e.g., the canonical chain and zero-or-many non-canonical forks).
- The cold database only contains finalized blocks and therefore it only needs to contain a single chain.

The "single chain" format of the cold database makes it much easier to optimise it for reduced disk usage and easy per-slot querying. Since there's a only a single chain we can lay things out linearly per-slot and de-duplicate things that are repeated in the beacon state (e.g., `state.block_roots`). We don't have this concept of linearity in the hot database and therefore it's harder for us to de-duplicate things and we can't key it per-slot because there might be multiple blocks in the same slot (i.e. forks).

So, the hot/cold separation gives us the ability to use two different schemas but it comes at the disk I/O cost of us having to migrate from hot -> cold as the split slot (i.e., finality) advances.

When it comes to blobs, we're still using the same schema between the hot and cold databases (i.e., blobs are keyed by their corresponding block root). This means we're not taking advantage of the different schemas but we're still paying the extra I/O cost of moving the blobs from the hot database into the cold database when the split slot advances.

So, rather than storing *only cold* blobs in a separate database, I propose that we store *all* blobs in a separate database. I believe this would save us the disk read/write cost of migrating from hot -> cold with all the benefits of the current approach.

My suggestion adds a little more complexity when `blobs_db_path = None`, though. With the current approach, we simply don't migrate to the cold DB. However with my suggestion, whenever store a blob in the DB we need to consider where to put the blob. I think we have three options when a user doesn't specify a `--blobs-db-path`:

1. Put blobs in the hot DB.
1. Put blobs in the cold DB.
1. Use a separate blobs DB in a default path (e.g. alongside the existing hot/cold DBs).

I'm a little uncertain here, I'll have a think and collect some feedback from people like @michaelsproul.

**Comment by paulhauner** [comment]
For the record, the current spec puts us at a maximum of ~68GB for blob storage:

- `MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS = 4096 epochs`
- `FIELD_ELEMENTS_PER_BLOB = 4096`
- `BYTES_PER_FIELD_ELEMENT = 32`
- `MAX_BLOBS_PER_BLOCK = 4`

`4096 * 32 * 4096 * 32 * 4 = 68,719,476,736`

This number doesn't account for:

- Another 88 bytes of block, slot and proof data in each `BlobsSidecar`.
   - At 4 blobs per block this is ~11MB: `((32 + 8 + 48)  * 4096 * 32) = 11,534,336`.
- We might not always have 4 blobs per block.

**Comment by emhane** [comment]
> I think the implementation here looks good, however I would like to suggest a conceptual change.
> 
> In this PR the concept of "hot" and "cold" blobs is introduced. A hot blob being one from a slot _later_ than the split slot (the split slot is approximates the finalized slot) and a cold blob being one from a slot _earlier_ than the split slot.
> 
> For context, one of the primary reasons we have the hot/cold database separation is because the DBs have different schemas due to the following differences:
> 
> * The hot database needs to be able to handle multiple chains (e.g., the canonical chain and zero-or-many non-canonical forks).
> * The cold database only contains finalized blocks and therefore it only needs to contain a single chain.
> 
> The "single chain" format of the cold database makes it much easier to optimise it for reduced disk usage and easy per-slot querying. Since there's a only a single chain we can lay things out linearly per-slot and de-duplicate things that are repeated in the beacon state (e.g., `state.block_roots`). We don't have this concept of linearity in the hot database and therefore it's harder for us to de-duplicate things and we can't key it per-slot because there might be multiple blocks in the same slot (i.e. forks).
> 
> So, the hot/cold separation gives us the ability to use two different schemas but it comes at the disk I/O cost of us having to migrate from hot -> cold as the split slot (i.e., finality) advances.
> 
> When it comes to blobs, we're still using the same schema between the hot and cold databases (i.e., blobs are keyed by their corresponding block root). This means we're not taking advantage of the different schemas but we're still paying the extra I/O cost of moving the blobs from the hot database into the cold database when the split slot advances.
> 
> So, rather than storing _only cold_ blobs in a separate database, I propose that we store _all_ blobs in a separate database. I believe this would save us the disk read/write cost of migrating from hot -> cold with all the benefits of the current approach.
> 
> My suggestion adds a little more complexity when `blobs_db_path = None`, though. With the current approach, we simply don't migrate to the cold DB. However with my suggestion, whenever store a blob in the DB we need to consider where to put the blob. I think we have three options when a user doesn't specify a `--blobs-db-path`:
> 
> 1. Put blobs in the hot DB.
> 2. Put blobs in the cold DB.
> 3. Use a separate blobs DB in a default path (e.g. alongside the existing hot/cold DBs).
> 
> I'm a little uncertain here, I'll have a think and collect some feedback from people like @michaelsproul.

I see your points, I was wondering if we wanted to make the blob smaller in any way like we do with state before storing in cold, that perhaps some *cryptography magic* could do this. But if you don't think an extra 68.7 + 0.046 GB isn't a problem, then I don't either.

Blobs in cold db is maybe better to not hold a lock on the hot db when pruning. What do you think?

**Comment by emhane** [comment]
> For the record, the current spec puts us at a maximum of ~68GB for blob storage:
> 
> * `MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS = 4096 epochs`
> * `FIELD_ELEMENTS_PER_BLOB = 4096`
> * `BYTES_PER_FIELD_ELEMENT = 32`
> * `MAX_BLOBS_PER_BLOCK = 4`
> 
> `4096 * 32 * 4096 * 32 * 4 = 68,719,476,736`
> 
Where is the second `32` coming from? Edit: Never mind, figured out it's the slots per epoch.
> This number doesn't account for:
> 
> * Another 88 bytes of block, slot and proof data in each `BlobsSidecar`.
>   
>   * At 4 blobs per block this is ~48MB: `((32 + 8 + 48) * 4 * 4096 * 32) = 46,137,344`.
> * We might not always have 4 blobs per block.

Each blobs sidecar holds all the blobs of a block so isn't it `((32 + 8 + 48)  * 4096 * 32) = 11,534,336` ?

**Comment by paulhauner** [comment]
> Each blobs sidecar holds all the blobs of a block so isn't it ((32 + 8 + 48)  * 4096 * 32) = 11,534,336 ?

Oh yep, you're right sorry!

File: beacon_node/store/src/hot_cold_store.rs
Comment (paulhauner) [inline beacon_node/store/src/hot_cold_store.rs]: I like how this block protects against the scenario where a user fails to provide a `blob-path` for a DB that's already been initialized with a separate blob DB. I can imagine the scenario where a user (wrongly) adds/removes the blobs DB flag for an existing database and ends up with LH complaining about missing blobs.

I wonder if we could also check the inverse scenario, where the DB already exists *without* a separate blobs DB but the users supplies a `blobs_db_path`. To achieve this I think we'd need to also store `BlobInfo { blobs_db: false }` when `blobs_db_path == None`. That we we could say:

- `db.load_blob_info()? == None` is a new database.
- `db.load_blob_info()? == Some(..)` is an existing database.

File: beacon_node/store/src/hot_cold_store.rs
Comment (paulhauner) [inline beacon_node/store/src/hot_cold_store.rs]: Yep, I think we delete the blob here too. I can't think of any circumstance where we'd want a block-less blob.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: I see that the `ops.push(StoreOp::PutBlobs(block_root, blobs))` method has been removed here and replaced with `self.store.put_blobs(&block_root, (&*blobs).clone())` on line 3004.

I also notice you've put some effort into the writing the [`do_atomically_with_block_and_blobs_cache`](https://github.com/sigp/lighthouse/blob/5eee07c33b87be7b8a9761d12fb1abc9d8b8ad26/beacon_node/store/src/hot_cold_store.rs#L825-L899) function, but I don't see where`StoreOp::PutBlobs` is constructed anymore.

I would expect this block of code (where I'm applying this comment) remains and the newly-added block containing line 3004 would be removed.

I could be missing something, though!

File: beacon_node/store/src/hot_cold_store.rs
Comment (paulhauner) [inline beacon_node/store/src/hot_cold_store.rs]: Are you asking if we should delete the blocks from the blob DB if the write fails? It seems like that would be the safest thing to do. However, I see that `do_atomically_with_block_and_blobs_cache` might also process other operations and reverting them all might get a little complex.

I would be tempted to try writing to the `blobs_db` first, then doing the `hot_db`. If the second write fails then I imagine it would be simpler to try and remove the blobs rather than revert the `hot_db` calls.

Perhaps @michaelsproul has some opinions on this DB atomicity decision?

**Review by paulhauner — APPROVED** [review]
Heyoo, I've managed to get to this review in transit! I'm presently stuck on the `StoreOp::PutBlobs` [question](https://github.com/sigp/lighthouse/pull/3869#discussion_r1091430981), could you please shed some light on that for me?

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (emhane) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Originally, blobs always went into the `hot_db` here, along with blocks, execution payload, etc. Since that line of code is not in a loop we are only putting one blobs sidecar in store. I knew we wanted to cache blobs so I used `put_blobs`. However, I wasn't sure if I we wanted to put blocks in cache here too, but here I got the answer https://github.com/sigp/lighthouse/pull/3931 and will use `do_atomically_with_block_and_blobs_cache`.

File: beacon_node/store/src/hot_cold_store.rs
Comment (emhane) [reply beacon_node/store/src/hot_cold_store.rs]: Yeah, that's exactly what I mean. Ok, that's a good idea to reverse the order and delete blobs instead.

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looks good on the whole. I think we should merge this soon so we can minimise the number of breaking DB changes (this PR's `BlobInfo` is incompatible with the one from `Prune blobs`).

I'll do a final pass once the merge conflicts are resolved

File: beacon_node/http_api/src/block_id.rs
Comment (realbigsean) [inline beacon_node/http_api/src/block_id.rs:224]: this is cool syntax i didn't know you could do this

**Review by realbigsean — APPROVED** [review]
Looks good to me, happy to merge it after Michael gives it a final look

File: beacon_node/http_api/src/block_id.rs
Comment (emhane) [reply beacon_node/http_api/src/block_id.rs:224]: 1.67 goodies. I keep up to date on every new release with these threads, I like the format, it's nice and easy to absorb the updates with the syntax highlighted screenshots included https://twitter.com/m_ou_se/status/1618635059455758336?s=46&t=0DikpsR8Q6qfdD_5KfxcFQ

**Review by michaelsproul — APPROVED** [review]
LGTM, lets try this on some 4844 devnets

**Pull request merged**
